File: exercises/python/lab04/fcfs.py
"""Lab 04C - First Come First Serve (FCFS) Scheduling"""

def fcfs(processes, burst_times):
    n = len(processes)
    wt  = [0] * n
    tat = [0] * n

    for i in range(1, n):
        wt[i] = wt[i-1] + burst_times[i-1]
    for i in range(n):
        tat[i] = burst_times[i] + wt[i]

    print("=== First Come First Serve (FCFS) Scheduling ===\n")
    print(f"{'Process':<12}{'Burst Time':<15}{'Waiting Time':<15}{'Turnaround Time'}")
    for i in range(n):
        print(f"  P{processes[i]:<10}{burst_times[i]:<15}{wt[i]:<15}{tat[i]}")

    print(f"\nAverage Waiting Time    = {sum(wt)/n:.2f}")
    print(f"Average Turnaround Time = {sum(tat)/n:.2f}")


if __name__ == "__main__":
    processes   = [1, 2, 3]
    burst_times = [10, 5, 8]
    fcfs(processes, burst_times)

File: exercises/cpp/lab04/fcfs.cpp
#include <iostream>
#include <vector>
using namespace std;

void findWaitingTime(int n, vector<int>& bt, vector<int>& wt) {
    wt[0] = 0;
    for (int i = 1; i < n; i++)
        wt[i] = bt[i-1] + wt[i-1];
}

void findTurnAroundTime(int n, vector<int>& bt, vector<int>& wt, vector<int>& tat) {
    for (int i = 0; i < n; i++)
        tat[i] = bt[i] + wt[i];
}

int main() {
    vector<int> processes = {1, 2, 3};
    vector<int> burst_time = {10, 5, 8};
    int n = processes.size();

    vector<int> wt(n), tat(n);
    findWaitingTime(n, burst_time, wt);
    findTurnAroundTime(n, burst_time, wt, tat);

    int total_wt = 0, total_tat = 0;
    cout << "=== First Come First Serve (FCFS) Scheduling ===\n\n";
    cout << "Process\t\tBurst Time\tWaiting Time\tTurnaround Time\n";
    for (int i = 0; i < n; i++) {
        total_wt  += wt[i];
        total_tat += tat[i];
        cout << "  P" << processes[i] << "\t\t" << burst_time[i]
             << "\t\t" << wt[i] << "\t\t" << tat[i] << "\n";
    }
    cout << "\nAverage Waiting Time    = " << (float)total_wt  / n << "\n";
    cout << "Average Turnaround Time = " << (float)total_tat / n << "\n";
    return 0;
}
